src/install.rs

```diff
@@ -345,56 +331,75 @@ fn write_platform(mountpoint: &Path, platform: &str) -> Result<()> {
-/// only install from metal images and that the bootloader configs will always set
-/// ignition.platform.id.  Fail if those assumptions change.  This is deliberately simplistic.
-fn bls_entry_write_platform(orig_contents: &str, platform: &str) -> Result<String> {
-    let new_contents = orig_contents.replace(
+/// To be used with `visit_bls_entry_options()`. Modifies the BLS config, only changing the
+/// `ignition.platform.id`. This assumes that we will only install from metal images and that the
+/// bootloader configs will always set ignition.platform.id.  Fail if those assumptions change.
+/// This is deliberately simplistic.
+fn bls_entry_write_platform(orig_options: &str, platform: &str) -> Result<Option<String>> {
+    let new_options = orig_options.replace(
         "ignition.platform.id=metal",
         &format!("ignition.platform.id={}", platform),
     );
-    if orig_contents == new_contents {
+    if orig_options == new_options {
         bail!("Couldn't locate platform ID");
     }
-    Ok(new_contents)
+    Ok(Some(new_options))
 }
 
-/// Apply a transforming function on each BLS entry found.
-pub fn edit_bls_entries(mountpoint: &Path, f: impl Fn(&str) -> Result<String>) -> Result<()> {
+/// Calls a function on the BLS entry and optionally updates it if the function returns new
+/// content. Errors out if none or more than one BLS entry was found.
+pub fn visit_bls_entry(
+    mountpoint: &Path,
+    f: impl Fn(&str) -> Result<Option<String>>,
+) -> Result<()> {
     // walk /boot/loader/entries/*.conf
     let mut config_path = mountpoint.to_path_buf();
     config_path.push("loader/entries");
+    let mut found_path: Option<PathBuf> = None;
     for entry in read_dir(&config_path)
         .chain_err(|| format!("reading directory {}", config_path.display()))?
     {
         let path = entry
             .chain_err(|| format!("reading directory {}", config_path.display()))?
             .path();
-        if path.extension().unwrap_or_default() == "conf" {
-            // slurp in the file
-            let mut config = OpenOptions::new()
-                .read(true)
-                .write(true)
-                .open(&path)
-                .chain_err(|| format!("opening bootloader config {}", path.display()))?;
-            let orig_contents = {
-                let mut s = String::new();
-                config
-                    .read_to_string(&mut s)
-                    .chain_err(|| format!("reading {}", path.display()))?;
-                s
-            };
-
-            let new_contents =
-                f(&orig_contents).chain_err(|| format!("modifying {}", path.display()))?;
+        if path.extension().unwrap_or_default() != "conf" {
+            continue;
+        }
+
+        if let Some(previous_path) = found_path {
+            bail!(
+                "Found more than one BLS entry: {} and {}",
+                previous_path.display(),
+                path.display()
+            );
+        }
```

Comment:
Do we actually need to bail if there is more than one BLS entry? Trying to use this in the upgrade script where we'll always have more than 1 BLS entry.

Reply:
Ouch, yes good catch. In basically all our cases *except* this one we really expect a single BLS entry. Rather than making it configurable, I just relaxed it so we don't really care how many BLS entries there. But instead, we make sure to only act on the latest/default BLS entry, which is what we want for `rdcore kargs`.

Reply:
In the case of the RAID conversion, any boot entries we don't rewrite will no longer be bootable.  Are we okay with that, or should we think about rewriting all of the entries?

Reply:
Do we support doing the RAID conversion post-firstboot, though?

Reply:
Sorry, I meant to say the RHCOS LUKS conversion.  And yes, that will only happen post-firstboot.

Reply:
Ahhh OK. Hmm, I lean towards accepting that risk. We don't even know if the previous deployments could even support the changes by just applying the same kargs (e.g. a user could've pinned a 6 month old deployment without necessary patches!). The only deployment we know supports the new configuration is the one we're running from, since it contains the code driving the modification. So IMO it only really makes sense to modify that deployment's BLS config.

(So actually, one thing worth pointing out here is that `rdcore kargs` assumes that you're booted in the latest/default BLS entry, because that's the one it's modifying. If a human does an `rpm-ostree deploy` and then `rpm-ostree rollback` to put the new OSTree with migration code in second place, rebooting into that second entry by manually selecting it at the boot menu will affect the wrong BLS config... we could catch that I think, though not sure how much we should care about that.)

Reply:
1. Okay, I'm okay with that.
2. Oof.  Well, we'd be affecting a BLS config other than the one we're booting, but is that actually the "wrong" config?  The user has opted to manually boot from a non-current config, but arguably that doesn't mean we should switch to maintaining an older one.

Reply:
It's the "wrong" config in the sense that we don't even know if the modifications we're making will work with the default deployment. But the whole problem is mainly theoretical; I don't think it's reasonable for migration code to handle the case where they're not the default config.